### PR TITLE
fix(common): canceled JSONP requests won't throw error with missing callback function

### LINKED ISSUE #34818: Canceling an in-progress JSONP request causes "callback undefined" console error
# 🐞 bug report

### Affected Package
@angular/common/http

### Is this a regression?
No

### Description

It looks like HttpClientJsonpModule deletes the JSONP callback function for a request immediately upon cancellation. If the request was still in progress, an error is logged in the console when it completes because the function doesn't exist.

This issue occurs when the time between requests is usually lower than the time a request takes to complete.

## 🔬 Minimal Reproduction
https://angular-c6oeqi.stackblitz.io
https://stackblitz.com/edit/angular-c6oeqi

## 🔥 Exception or Error
<pre><code>
Uncaught ReferenceError: ng_jsonp_callback_16 is not defined
</code></pre>

## 🌍  Your Environment
_**Can be reproduced with Angular 8 (see repro)**_ ‼️

<pre><code>
Angular CLI: 6.2.9
Node: 8.17.0
OS: darwin x64
Angular: 6.1.10
... animations, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic
... platform-server, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.8.9
@angular-devkit/build-angular     0.8.9
@angular-devkit/build-optimizer   0.8.9
@angular-devkit/build-webpack     0.8.9
@angular-devkit/core              0.8.9
@angular-devkit/schematics        0.8.9
@angular/cdk                      6.4.7
@angular/cli                      6.2.9
@angular/language-service         7.2.15
@ngtools/webpack                  6.2.9
@schematics/angular               0.8.9
@schematics/update                0.8.9
rxjs                              6.2.2
typescript                        2.9.2
webpack                           4.16.4
</code></pre>

**Anything else relevant?**
No

### PR DESCRIPTION
**Pull request by martinsik**
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Fixes #34818

Unsubscribing from a JSONP request will throw console errors such as `Uncaught ReferenceError: ng_jsonp_callback_41 is not defined`. This is happening because the callback function is removed right on unsubscription but this doesn't stop the JSONP script from loading.

https://stackblitz.com/edit/angular-4zmfjr?file=src/app/app.component.ts

## What is the new behavior?

Unsubscribing from JSONP request won't remove callback handler. Handler is then removed by the handler itself:

https://github.com/angular/angular/blob/2365bb89d7a37089f596705db19b1ee5d0366b42/packages/common/http/src/jsonp.ts#L107

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/http/test/jsonp_spec.ts [hunk 1]
```diff
@@ -18,7 +18,6 @@ function runOnlyCallback(home: any, data: Object) {
   const keys = Object.keys(home);
   expect(keys.length).toBe(1);
   const callback = home[keys[0]];
-  delete home[keys[0]];
```
Comment (martinsik) [inline packages/common/http/test/jsonp_spec.ts:21]: I removed this line because the handler should be removed by the handler itself:

https://github.com/angular/angular/blob/2365bb89d7a37089f596705db19b1ee5d0366b42/packages/common/http/src/jsonp.ts#L107

**Review by JoostK — COMMENTED** [review]
It appears to be possible to prevent the script from executing by adopting the script element into another document, as described in https://html.spec.whatwg.org/multipage/scripting.html#execute-the-script-block. So instead of removing the script element as we do now, it could instead be adopted into a dummy document. I think that doing so is beneficial in terms of resources, as it avoids the need to parse and execute the script.

**Comment by petebacondarwin** [comment]
@martinsik - would you have time to address the comments from @JoostK?

**Comment by martinsik** [comment]
It's been so long I've completelly forgot I ever made this PR :).

@JoostK I made a small demo to test if changing <script>'s document will prevent it from exeuction and it does work https://stackblitz.com/edit/js-gs6nph?file=index.js  
I refactored the code a little bit. Seems like one state variable (`cancelled`) could be removed completely and I shouldn't even need to use `removeEventListener` because the events won't be triggered after changing the owner document.

I'm not sure why `ci/circleci: legacy-unit-tests-saucelab`  CICD process is failing but it looks like the same problem is in other PRs as well so it doesn't look like I broke it.

File: packages/common/http/test/jsonp_spec.ts [hunk 2]
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
       });
       document.mockError(error);
     });
+    it('allows the callback to be invoked when the request is cancelled', () => {
+      backend.handle(SAMPLE_REQ).subscribe().unsubscribe();
+      expect(Object.keys(home).length).toBe(0);
+      expect(document.mock!.ownerDocument).not.toEqual(document);
```
Comment (JoostK) [inline packages/common/http/test/jsonp_spec.ts:68]: Please add a comment here to indicate why this assertion is relevant. Something like:

```ts
// The script element should have been transferred to a different document to prevent it from executing.
```

File: packages/common/http/test/jsonp_spec.ts [hunk 3]
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
       });
       document.mockError(error);
     });
+    it('allows the callback to be invoked when the request is cancelled', () => {
+      backend.handle(SAMPLE_REQ).subscribe().unsubscribe();
+      expect(Object.keys(home).length).toBe(0);
```
Comment (JoostK) [inline packages/common/http/test/jsonp_spec.ts:67]: Is this sufficient to verify that the callback handler wasn't executed? Isn't this check only indicating that cleanup happened?

File: packages/common/http/test/jsonp_spec.ts [hunk 4]
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
       });
       document.mockError(error);
     });
+    it('allows the callback to be invoked when the request is cancelled', () => {
```
Comment (JoostK) [inline packages/common/http/test/jsonp_spec.ts:65]: I remember proposing this spec description and looking at it now it no longer makes sense to me (sorry!), as there is no sign of a callback anywhere in this test setup, and I think the callback should not at all be invoked, right?

So here's attempt two for a description:

```ts
    it('prevents the script from executing when the request is cancelled', () => {
```

would that be better?

**Review by JoostK — COMMENTED** [review]
Thanks for your continued effort on this PR. It looks good to me; I'll request a review from @alxhub as code owner of fw-http so we can get this landed.

**Comment by JoostK** [comment]
Btw don't worry about the Saucelabs failures; they have been _very_ flaky over the last couple of weeks and #44281 should help make it more stable.

File: packages/common/http/test/jsonp_spec.ts (hunk 3 above)
Comment (martinsik) [reply packages/common/http/test/jsonp_spec.ts:73]: I updated the test to also check that the callback wasn't invoked.

**Comment by martinsik** [comment]
@JoostK I updated this PR and implemented your comments.

File: packages/common/http/src/jsonp.ts [hunk 5]
```diff
@@ -56,6 +56,8 @@ export class JsonpClientBackend implements HttpBackend {
    */
   private readonly resolvedPromise = Promise.resolve();
 
+  private readonly dummyDocument = this.document.implementation.createHTMLDocument();
+
```
[the remark this answers is not in the document]
Comment (martinsik) [reply packages/common/http/src/jsonp.ts]: @AndrewKushnir I moved it into a separate method. I'm not sure if `removeListeners` isn't confusing because it's not really removing any listeners.

**Comment by martinsik** [comment]
@AndrewKushnir I rebased the code and implemented your suggestions.

**Review by AndrewKushnir — APPROVED** [review]
@martinsik the change looks good 👍 

I'd like to ask @JoostK to take another look. After that we can run the necessary tests in Google's codebase and proceed with next steps if everything looks good.

Thank you.

File: packages/common/http/src/jsonp.ts [hunk 6]
```diff
@@ -56,6 +56,12 @@ export class JsonpClientBackend implements HttpBackend {
    */
   private readonly resolvedPromise = Promise.resolve();
 
+  /**
+   * When a pending <script> is unsubscribed we'll move it to this document, so it won't be
+   * executed.
+   */
+  private foreignDocument!: Document;
```
Comment (JoostK) [inline packages/common/http/src/jsonp.ts:63]: nit: I'd personally move this to become a top-level variable, as this can be safely shared between multiple instances (although unlikely to be relevant except for tests) and it minimizes better.

I'd also request not to use `!` here, as we can easily include `|undefined` in the type to make reads of this field null-safe. Or use `Document|null = null;` which produces more code but has the benefit that reads from this class stay monomorphic. But that wouldn't be a concern if it's a top-level var (nor is this performance sensitive code anyway, I'm mentioning it as a general takeaway).

tldr; I'd declare `let foreignDocument: Document|undefined;` as top-level declaration.

**Review by JoostK — APPROVED** [review]

File: packages/common/http/src/jsonp.ts (hunk 6 above)
Comment (martinsik) [reply packages/common/http/src/jsonp.ts]: @JoostK I moved `foreignDocument` outside of the class, but it seems like I'll need to be using `!` in `foreignDocument!.adoptNode(script);` because `this.document` is of type `any` so TypeScript can't know what type is returned from `createHTMLDocument()` .
I could use typecasting on this `(this.document.implementation as DOMImplementation).createHTMLDocument()` but that's probably not correct because `document` could use a different implementation.

Comment (JoostK) [reply packages/common/http/src/jsonp.ts]: Ah, interesting. If we're relying on `createHTMLDocument` to be available, then we might as well do the cast. If we don't want to rely on that then we'd need to account for the situation where `implementaiton` or `createHTMLDocument` is not available as we'd expect, but I don't think there's a reasonable alternative we could use in that case so that seems not viable to me.

Comment (martinsik) [reply packages/common/http/src/jsonp.ts]: Ok, I added typecast to `DOMImplementation`.

**Comment by JoostK** [comment]
@martinsik thanks, LGTM. Could you please rebase on latest master to get CI to run?

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 909b21aa29c1fe8094ce3f7fb70ddf24668ba155.
